### PR TITLE
[core] Throw SnapshotNotExistException when the snapshot for creating tag does not exist.

### PR DESCRIPTION
**Pull request by LinMingQiang**
### Purpose

Creating failed : 
image

Creating successful : 
image

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java [hunk 1]
```diff
@@ -61,25 +67,45 @@ public String[] call(
...
+        TagManager tagManager = fileStoreTable.tagManager();
+        Snapshot fromSnapshot = null;
         if (snapshotId == null) {
-            table.createTag(tagName, toDuration(timeRetained));
+            fromSnapshot = snapshotManager.latestSnapshot();
         } else {
-            table.createTag(tagName, snapshotId, toDuration(timeRetained));
+            try {
+                fromSnapshot = tagManager.createTagInternal(snapshotId, snapshotManager);
+            } catch (Exception ignored) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java:92]: Here should not catch all exceptions.

Maybe we should refactor `createTagInternal` return value or throw a special exception.

Comment (LinMingQiang) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java:92]: Just catch `IllegalArgumentException`.

Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java]: You could create a new Exception class. For example, `SnapshotNotExistException`

**Comment by LinMingQiang** [comment]
flaky test : https://github.com/apache/paimon/pull/3960

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/CreateTagProcedure.java [hunk 2]
```diff
@@ -47,7 +52,7 @@ public class CreateTagProcedure extends BaseProcedure {
     private static final StructType OUTPUT_TYPE =
             new StructType(
                     new StructField[] {
-                        new StructField("result", DataTypes.BooleanType, true, Metadata.empty())
+                        new StructField("snapshot", StringType, true, Metadata.empty())
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/CreateTagProcedure.java:55]: why is snapshot? not result?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java [hunk 3]
```diff
@@ -61,25 +68,45 @@ public String[] call(
...
 
-    public String[] call(
+    public @DataTypeHint("ROW<snapshot STRING>") Row[] call(
             ProcedureContext procedureContext, String tableId, String tagName, String timeRetained)
             throws Catalog.TableNotExistException {
         return innerCall(tableId, tagName, null, timeRetained);
     }
 
-    private String[] innerCall(
+    private @DataTypeHint("ROW<snapshot STRING>") Row[] innerCall(
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java:77]: ditto

Comment (LinMingQiang) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CreateTagProcedure.java]: fixed.

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/CreateTagProcedure.java [hunk 4]
```diff
@@ -75,13 +80,38 @@ public InternalRow[] call(InternalRow args) {
...
+                                                    "Cannot create tag because the %s doesn't exist.",
+                                                    snapshot == null
+                                                            ? "latest snapshot"
+                                                            : "snapshot #" + snapshot)))
+                        };
                     }
-                    InternalRow outputRow = newInternalRow(true);
-                    return new InternalRow[] {outputRow};
+
+                    fileStoreTable.createTag(tag, fromSnapshot.id(), timeRetained);
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/CreateTagProcedure.java:110]: Why create tag again?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**